ch-image: add export (#1515)
hpc/charliecloud: doc/ch-image.rst

   $ ch-image [...] build [-t TAG] [-f DOCKERFILE] [...] CONTEXT
   $ ch-image [...] build-cache [...]
   $ ch-image [...] delete IMAGE_REF

>    $ ch-image [...] export [-o FMT] [...] IMAGE_REF IMAGE_OUT

Reviewer: ```suggestion
   $ ch-image [...] export [-o FMT] [...] IMAGE_REF PATH
```

   Glob characters must be quoted or otherwise protected from the shell, which
   also desires to interpret them and will do so incorrectly.

:code:`export`
===============

::

   $ ch-image [...] export [-o FMT] [...] IMAGE_REF PATH

Description
-----------

> Copy image :code:`IMAGE_REF` to :code:`PATH` and convert its format. Replace
> :code:`PATH` if it already exists, unless :code:`--no-clobber` is specified. It
> is an error if :code:`IMAGE_REF` and :code:`PATH` have the same format; use the
> format's own tools for that case.

Reviewer: ```suggestion
Copy image :code:`IMAGE_REF` from builder storage to :code:`PATH` in format
:code:`FMT` if provided; otherwise, use default format :code:`tar`. Replace
image at :code:`PATH` if it already exists, unless :code:`--no-clobber` is
specified.
```

> :code:`ch-run` can run container images that are plain directories or
> (optionally) SquashFS archives. However, images can take on a variety of other
> formats as well. The main purpose of this tool is to make images in those
> other formats available to :code:`ch-run`.

Reviewer: ```suggestion
:code:`ch-run` can run container images that are plain directories or
(optionally) SquashFS archives. However, images can take on a variety of other
formats as well. The purpose of this tool is to export images from builder
storage to a supported image format.
```

For best performance, :code:`export` should be invoked only once,
producing the final format actually needed.

>   :code:`IMAGE_REF`
>     Descriptor for the input image. For image builders, this is an image
>     reference; otherwise, it's a filesystem path.

Reviewer: ```suggestion
  :code:`IMAGE_REF`
    Image in builder storage to be exported.
```

  :code:`PATH`

>     Descriptor for the output image.

Reviewer: ```suggestion
    Destination path for exported image.
```

>   :code:`-h`, :code:`--help`
>     Print help and exit.

Reviewer: ```suggestion
```

Reviewer: Help info is stated elsewhere by argparse.

  :code:`-n`, :code:`--dry-run`
    Don't read the input or write the output. Useful for testing format

>     inference.

Reviewer: We should discuss whether this subcommand should infer like `ch-convert` does offline with @reidpr and friends.

>   :code:`--no-clobber`
>     Error if :code:`OUT` already exists, rather than replacing it.

Reviewer: ```suggestion
  :code:`--no-clobber`
    Error if image at :code:`PATH` already exists, rather than replacing it.
```

  :code:`-o`, :code:`--out-fmt FMT`
    Output image format is :code:`FMT`; inferred if omitted.

  :code:`--tmp DIR`

>     A sub-directory for temporary storage is created in :code:`DIR` and
>     removed at the end of a successful conversion. **If this script crashes or
>     errors out, the temporary directory is left behind to assist in
>     debugging.** Storage may be needed up to twice the uncompressed size of
>     the image, depending on the input and output formats. Default:
>     :code:`$TMPDIR` if specified; otherwise :code:`/var/tmp`.

Reviewer: ```suggestion
    A sub-directory for temporary storage is created in :code:`DIR` and
    removed at the end of a successful conversion. **If this subcommand
    crashes or errors out, the temporary directory is left behind to assist in
    debugging.** Storage may be needed up to twice the uncompressed size of
    the image, depending on the export output format. Default:
    :code:`$TMPDIR` if specified; otherwise :code:`/var/tmp`.
```

> .. Notes:
>    1. It's a deliberate choice to use UNIXey options rather than the Skopeo
>       syntax [1], e.g. "-i docker" rather than "docker:image-name".
>       [1]: https://manpages.debian.org/unstable/golang-github-containers-image/containers-transports.5.en.html

Reviewer: ```suggestion
```

Other builders are irrelevant.

>    2. There used to be an [OUT_ARG ...] that would be passed unchanged to the
>       archiver, i.e. tar(1) or mksquashfs(1). However it wasn't clear there
>       were real use cases, and this has lots of opportunities to mess things
>       up. Also, it's not clear when it will be called. For example, if you
>       convert a directory to a tarball, then passing e.g. -J to XZ-compress
>       will work fine, but when converting from Docker, we just compress the
>       tarball we got from Docker, so in that case -J wouldn't work.
>    3. I also deliberately left out an option to change the output compression
>       algorithm, under the assumption that the default is good enough. This
>       can be revisited later IMO if needed.

Reviewer: ```suggestion
```
Remove for now. We can revise after discussing.

Image formats
-------------

:code:`export` knows about these values of :code:`FMT`:

>   :code:`ch-image`
>     Internal storage for Charliecloud's unprivileged image builder (Dockerfile
>     interpreter) :code:`ch-image`.

Reviewer: ```suggestion
```

Remove. This is implied.

  :code:`dir`
    Ordinary filesystem directory (i.e., not a mount point) containing an
    unpacked image. Output directories that already exist are replaced if they

>     look like an image; otherwise, exit with an error.

Reviewer: ```suggestion
    look like an image, unless :code:`--no-clobber` is
    specified; otherwise, exit with an error.
```

  :code:`squash`
    SquashFS filesystem archive containing the flattened image. SquashFS
    archives are much like tar archives but are mountable, including by
    :code:`ch-run`'s internal SquashFUSE mounting. Most systems have at least
    the SquashFS-Tools installed which allows unpacking into a directory, just
    like tar. Due to this greater flexibility, SquashFS is preferred to tar.

>     **Note:** Conversions to and from SquashFS are quite noisy due to the
>     verbosity of the underlying :code:`mksquashfs(1)` and
>     :code:`unsquashfs(1)` tools.

Reviewer: ```suggestion
    **Note:** Conversions to SquashFS are quite noisy due to the verbosity
    of the underlying :code:`mksquashfs(1)` and :code:`unsquashfs(1)` tools.
```

  :code:`tar`
    Tar archive containing the flattened image with no layer sub-archives.
    Output tarballs are always gzipped and must end in :code:`.tar.gz`;
    input tarballs can be any compression acceptable to :code:`tar(1)`.

All of these are local formats; :code:`export` does not know how to push
or pull images.

Format inference
----------------

:code:`export` tries to save typing by guessing formats when they are
reasonably clear. This is done against filenames, rather than file contents,
so the rules are the same for output descriptors that do not yet exist.

Format inference is done for both :code:`IN` and :code:`OUT`. The first
matching glob below yields the inferred format. Paths need not exist in the
filesystem.

  1. :code:`*.sqfs`, :code:`*.squash`, :code:`*.squashfs`: SquashFS.

  2. :code:`*.tar`, :code:`*.t?z`, :code:`*.tar.?`, :code:`*.tar.??`: Tarball.

  3. :code:`/*`, :code:`./*`, i.e. absolute path or relative path with
     explicit dot: Directory.

  4. If `ch-image` is installed: :code:`ch-image` internal storage.

>   5. Otherwise: No format inference.

Reviewer: See previous inference uncertainty.

Author: Format inference should have the functionality of ch-convert's format inference

Examples
--------

Typical build-to-run sequence for image :code:`foo/bar` using :code:`ch-run`'s
internal SquashFUSE code, inferring the output format::

  $ ch-image build -t hello .
  [...]

>   $ ch-image export hello /var/tmp/hello.sqfs

Reviewer: ```suggestion
  $ ch-image export hello /var/tmp
```

I think, at least as it reads, that `PATH` doesn't include the image name. For example, `
ch-image export hello:latest /tmp -o squash` will result in a file `hello:latest.sqfs` at `/tmp`, e.g., `/tmp/hello:latest.sqfs`

  input:   ch-image  hello
  output:  squashfs  /var/tmp/hello.sqfs
  copying ...
  done
  $ ch-run /var/tmp/hello.sqfs -- echo hello
  hello

Same conversion, but no format inference::

  $ ch-image export -o squash hello /var/tmp/foobar.sqfs
  input:   ch-image  hello
  output:  squashfs  /var/tmp/hello.sqfs
  copying ...
  done

> 

Reviewer: Ah, we will want to update this to also include an example covering @reidpr's suggest supplemental build argument, `--output`, e.g., `$ch-image build --output [...]`.
